t/t1092-sparse-checkout-compatibility.sh

```diff
@@ -588,6 +588,144 @@ test_expect_success 'reset with wildcard pathspec' '
+	test_all_match git update-index --force-remove --ignore-skip-worktree-entries folder1/a &&
+	test_all_match git status --porcelain=v2
+'
+
+test_expect_success 'update-index folder add/remove' '
+	init_repos &&
+
+	test_all_match test_must_fail git update-index --add --remove deep &&
+	test_all_match test_must_fail git update-index --add --remove deep/ &&
+
+	# NEEDSWORK: attempting to update-index on an existing folder outside the
+	# sparse checkout definition does not throw an error (as it does for folders
+	# inside the definition, or in the full checkout). However, it is a no-op.
+	test_sparse_match git update-index --add --remove folder1 &&
+	test_sparse_match git update-index --add --remove folder1/ &&
+	test_sparse_match git update-index --force-remove folder1/ &&
+	test_all_match git status --porcelain=v2 &&
+
+	# New folders, even in sparse checkouts, throw an error on update-index
+	run_on_all mkdir folder3 &&
+	run_on_all cp a folder3/a &&
+	run_on_all test_must_fail git update-index --add --remove folder3
+'
+
+test_expect_success 'update-index with updated flags' '
+	init_repos &&
+
+	# NEEDSWORK: updating flags runs inconsistently on directories, performing no
+	# operation with warning text specifying the path being ignored if a trailing
+	# slash is in the path, but throwing an error if there is no trailing slash.
+	test_all_match test_must_fail git update-index --no-skip-worktree folder1 &&
+	test_all_match git update-index --no-skip-worktree folder1/ &&
+	test_all_match git status --porcelain=v2 &&
+
+	# Removing the skip-worktree bit from a file outside the sparse checkout
+	# will cause the file to appear as unstaged and deleted.
+	test_sparse_match git update-index --no-skip-worktree folder1/a &&
+	rm -f full-checkout/folder1/a &&
+	test_all_match git status --porcelain=v2
+'
+
+test_expect_success 'update-index --again' '
+	init_repos &&
+
+	write_script edit-contents <<-\EOF &&
+	echo text >>$1
+	EOF
+
+	# With no local changes (staged or unstaged), update-index --again behaves
+	# consistently across checkouts
+	test_all_match git update-index --remove --again &&
+	test_all_match git status --porcelain=v2 &&
+
+	run_on_all ../edit-contents deep/README.md &&
+	test_all_match git update-index --add deep/README.md &&
+	run_on_all rm -f deep/README.md &&
+	test_all_match test_must_fail git update-index --again &&
+	test_all_match git update-index --remove --again &&
+
+	# Reupdate operates only on modified entries in the cache, so if a cache
+	# entry is present (inside or outside of the sparse checkout definition),
+	# it will be re-updated.
+	run_on_sparse mkdir -p folder1 &&
+	run_on_all ../edit-contents folder1/a &&
+	test_sparse_match git update-index --no-skip-worktree folder1/a &&
+	test_all_match git update-index --add folder1/a &&
+
+	run_on_all rm -f folder1/a &&
+	test_all_match test_must_fail git update-index --again folder1 &&
+	test_all_match git update-index --remove --again
+'
```

Comment:
What happens here if we add the following?

```
	run_on_sparse mkdir folder1 &&
	run_on_all ../edit-contents folder1/a &&
	test_all_match git update-index --again
```

(This `test_all_match` might need to be `test_sparse_match`, but worth a shot to explore.)

Generally, these cases of an external actor adding files outside of the sparse-checkout cone is not very well-tested in the Git test suite, but these cases are the most likely way to break the sparse index.

Reply:
Due to the last test (`git update-index --remove --again`), `folder1/a` will be removed from the index, staged for commit as "deleted". When `folder1/a` is added back on-disk, it is shown as an untracked file across all three checkouts and `update-index --again` (even with `--add`) does nothing.

Would you like that case added to the tests? Or was there a different result you were looking for (that I could try to replicate some other way)?

Reply:
> Due to the last test (`git update-index --remove --again`), `folder1/a` will be removed from the index, staged for commit as "deleted". When `folder1/a` is added back on-disk, it is shown as an untracked file across all three checkouts and `update-index --again` (even with `--add`) does nothing.

That's what happens when `folder1` is missing, as expected in a sparse-checkout. The test I'm recommending is about what happens when `folder1/a` exists.

> Would you like that case added to the tests? Or was there a different result you were looking for (that I could try to replicate some other way)?

I'm only recommending that we explore the mode where users or other tools recreate files outside the sparse-checkout cone. See the `'add outside sparse cone'` test for some examples. These are rare corner cases, but they can happen, and it's the most likely area where we will unintentionally create bugs if we don't add the tests now.

Reply:
> That's what happens when `folder1` is missing, as expected in a sparse-checkout.

That's not quite what I meant - during the test, the file _is_ manually added to the sparse checkouts (and `skip-worktree` is disabled, so its changes are shown in the index). It is manually deleted in the test I was referencing after all of that, so its deletion is tracked.

In any case, my comment earlier describes the cases added/modified - hopefully that covers what you're looking for: if the file is currently skipped with `skip-worktree`, on disk or otherwise, it's ignored; if it's _not_ ignored (because `skip-worktree` is ignored), everything behaves the same as in `full-checkout`.

Reply:
> > That's what happens when `folder1` is missing, as expected in a sparse-checkout.
> 
> That's not quite what I meant - during the test, the file _is_ manually added to the sparse checkouts (and `skip-worktree` is disabled, so its changes are shown in the index). It is manually deleted later on in the test, so its deletion is tracked because of that.

Ah, I misread the situation. Thanks for clarifying. That also means the file is not within a sparse directory entry, because the flipped `SKIP_WORKTREE` bit prevents that.

> In any case, my comment earlier describes the cases added/modified - hopefully that covers what you're looking for: if the file is currently skipped with `skip-worktree`, on disk or otherwise, it's ignored; if it's _not_ ignored (because `skip-worktree` is ignored), everything behaves the same as in `full-checkout`.

The biggest scenario I'm trying to understand is what happens when the index stores a sparse directory (such as `folder1/`) but the working tree has a file within that directory (such as `folder1/a`) and the user asks to update the index to match what's in the working tree.

Reply:
> The biggest scenario I'm trying to understand is what happens when the index stores a sparse directory (such as `folder1/`) but the working tree has a file within that directory (such as `folder1/a`) and the user asks to update the index to match what's in the working tree.

I _think_ this is covered by `update-index add outside sparse definition`. In that test, `folder1/` is sparse, but `a` is manually added to the working copy. After running `git update-index --add folder1/a`, it's still not added to the index (unless `--no-ignore-skip-worktree-entries` is specified which, as you mentioned, forcibly changes the sparseness of the directory).

Reply:
The new test covers this case, thanks.